server/datastore/mysql/software.go

```diff
@@ -200,58 +202,136 @@ func insertNewInstalledHostSoftwareDB(
 	return nil
 }
 
-func listSoftwareDB(ctx context.Context, q sqlx.QueryerContext, hostID *uint, opt fleet.SoftwareListOptions) ([]fleet.Software, error) {
-	hostWhere := `hs.host_id=?`
-	if hostID == nil {
-		hostWhere = "TRUE"
-	}
-	teamWhere := `h.team_id=?`
-	if opt.TeamID == nil {
-		teamWhere = "TRUE"
-	}
-	vulnerableJoin := "LEFT JOIN software_cpe scp ON (s.id=scp.software_id)"
-	if opt.VulnerableOnly {
-		vulnerableJoin = `JOIN software_cpe scp ON (s.id=scp.software_id)
-		JOIN software_cve scv ON (scp.id=scv.cpe_id)`
-	}
-	sql := fmt.Sprintf(`
-		SELECT DISTINCT s.*, coalesce(scp.cpe, "") as generated_cpe
-		FROM host_software hs
-		JOIN hosts h ON (hs.host_id=h.id)
-		JOIN software s ON (hs.software_id=s.id)
-		%s
-		WHERE %s AND %s
-	`, vulnerableJoin, hostWhere, teamWhere)
+var dialect = goqu.Dialect("mysql")
+
+// listSoftwareDB returns all the software installed in the given hostID and list options.
+// If hostID is nil, then the method will look into the installed software of all hosts.
+func listSoftwareDB(
+	ctx context.Context, q sqlx.QueryerContext, hostID *uint, opts fleet.SoftwareListOptions,
+) ([]fleet.Software, error) {
+	ds := dialect.From(goqu.I("host_software").As("hs")).SelectDistinct(
+		"s.*",
+		goqu.COALESCE(goqu.I("scp.cpe"), "").As("generated_cpe"),
+	).Join(
+		goqu.I("hosts").As("h"),
+		goqu.On(
+			goqu.I("hs.host_id").Eq(goqu.I("h.id")),
+		),
+	).Join(
+		goqu.I("software").As("s"),
+		goqu.On(
+			goqu.I("hs.software_id").Eq(goqu.I("s.id")),
+		),
+	)
+
+	if opts.VulnerableOnly {
+		ds = ds.Join(
```

Comment:
So would we be able to add joins later on to make it more readable for situations like this https://github.com/fleetdm/fleet/pull/2753#discussion_r741498949 ?

Reply:
Confirmed, for example: we can move the `Join` calls to the very end, just before generating the `sql string` and it works (same query is generated).

Reply:
Awesome, let's do this! :)